### PR TITLE
First commits

### PR DESCRIPTION
**Pull request by jsf9k**
## 🗣 Description ##

This pull request contains the first commits for this Terraform module.

## 💭 Motivation and context ##

We require such a generic capability to avoid duplicated code for each and every one of our EC2 instances.

## 🧪 Testing ##

All automated tests pass.  This code is also already deployed to our staging COOL Shared Services account and it appears to be functioning as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,42 +43,52 @@ No modules.
...
+| disk\_utilization\_alarm\_parameters | An object containing the parameters for the disk utilization alarm (e.g. {create\_alarm = true, datapoints\_to\_alarm = 1, evaluation\_periods = 1, period = 60, statistic = "Maximum", threshold = 90.0}).  See [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html) for an explanation of the parameters.  The default is to alarm for greater then 90% disk utilization. | `object({ create_alarm = bool, datapoints_to_alarm = number, evaluation_periods = number, period = number, statistic = string, threshold = number })` | ```{ "create_alarm": true, "datapoints_to_alarm": 1, "evaluation_periods": 1, "period": 300, "statistic": "Maximum", "threshold": 90 }``` | no |
```
Comment (dav3r) [inline README.md]: Typo:
```suggestion
| disk\_utilization\_alarm\_parameters | An object containing the parameters for the disk utilization alarm (e.g. {create\_alarm = true, datapoints\_to\_alarm = 1, evaluation\_periods = 1, period = 60, statistic = "Maximum", threshold = 90.0}).  See [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html) for an explanation of the parameters.  The default is to alarm for greater than 90% disk utilization. | `object({ create_alarm = bool, datapoints_to_alarm = number, evaluation_periods = number, period = number, statistic = string, threshold = number })` | ```{ "create_alarm": true, "datapoints_to_alarm": 1, "evaluation_periods": 1, "period": 300, "statistic": "Maximum", "threshold": 90 }``` | no |
```

File: variables.tf [hunk 2]
```diff
@@ -4,30 +4,48 @@
...
+  default     = { create_alarm = true, datapoints_to_alarm = 6, evaluation_periods = 6, period = 300, statistic = "Maximum", threshold = 90.0 }
 }
 
-variable "aws_region" {
-  type        = string
-  description = "The AWS region to deploy into (e.g. us-east-1)."
-  default     = "us-east-1"
+variable "disk_utilization_alarm_parameters" {
+  type        = object({ create_alarm = bool, datapoints_to_alarm = number, evaluation_periods = number, period = number, statistic = string, threshold = number })
+  description = "An object containing the parameters for the disk utilization alarm (e.g. {create_alarm = true, datapoints_to_alarm = 1, evaluation_periods = 1, period = 60, statistic = \"Maximum\", threshold = 90.0}).  See [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html) for an explanation of the parameters.  The default is to alarm for greater then 90% disk utilization."
```
Comment (dav3r) [inline variables.tf]: Typo:
```suggestion
  description = "An object containing the parameters for the disk utilization alarm (e.g. {create_alarm = true, datapoints_to_alarm = 1, evaluation_periods = 1, period = 60, statistic = \"Maximum\", threshold = 90.0}).  See [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html) for an explanation of the parameters.  The default is to alarm for greater than 90% disk utilization."
```

File: aws_ec2_alarms.tf [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      period      = 60
+      stat        = "Maximum"
+    }
+    return_data = true
+  }
+  ok_actions = var.ok_actions
+  threshold  = 0
+}
+
+# Alarm is an IMDSv1 request ever succeeds.
```
Comment (dav3r) [inline aws_ec2_alarms.tf]: Typo:
```suggestion
# Alarm if an IMDSv1 request ever succeeds.
```

File: cwagent_alarms.tf [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      period      = var.memory_utilization_alarm_parameters.period
+      stat        = var.memory_utilization_alarm_parameters.statistic
+    }
+    return_data = true
+  }
+  ok_actions = var.ok_actions
+  threshold  = var.memory_utilization_alarm_parameters.threshold
+}
+
+# Alarm for disk utilization
```
Comment (dav3r) [inline cwagent_alarms.tf]: Add period for consistency with other comments:
```suggestion
# Alarm for disk utilization.
```

**Review by dav3r — APPROVED** [review]
Outstanding - I have nothing to add functionally, just spotted a couple of typos.  👍 👍

File: README.md (hunk 1 above)
Comment (jsf9k) [reply README.md]: Please see commit 1c83547.

File: variables.tf (hunk 2 above)
Comment (jsf9k) [reply variables.tf]: Please see commit 1c83547.

File: aws_ec2_alarms.tf [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      stat        = "Maximum"
+    }
+    return_data = true
+  }
+  ok_actions = var.ok_actions
+  threshold  = 0
+}
+
+# Alarm for CPU utilization
+resource "aws_cloudwatch_metric_alarm" "cpu_utilization" {
```
Comment (mcdonnnj) [inline aws_ec2_alarms.tf:92]: I am confused on why CPU utilization is in this Terraform file but memory and disk utilization are in a different file. Based on alarm names the latter two should also be in this file.

**Review by mcdonnnj — APPROVED** [review]
Strong work 💪💪💪. I had one comment for your consideration.

File: aws_ec2_alarms.tf (hunk 5 above)
Comment (jsf9k) [reply aws_ec2_alarms.tf:92]: I put all the alarms based on metrics that are published to the "automatic" `AWS/EC2` metric namespace in `aws_ec2_alarms.tf`.  I put in `cwagent_alarms.tf` all the alarms based on "extra" metrics that are published by the CloudWatch Agent to the `CWAgent` metric namespace.

The CPU utilization happens to be in for former category, while the memory and disk metrics happen to be in the latter.

Comment (mcdonnnj) [reply aws_ec2_alarms.tf:92]: Ok, that makes sense. That distinction wasn't clear from how things are organized.

Comment (jsf9k) [reply aws_ec2_alarms.tf:92]: I can add/improve comments at the top of each of the two `*.tf` files.

Comment (jsf9k) [reply aws_ec2_alarms.tf:92]: Please see commit 8e0c2d8.

**Review by mcdonnnj — APPROVED** [review]
Thanks for the documentation addition! *Approval intensifies.*

**Pull request merged**